`common_messages_sv2` docs

jbesraa opened the pull request:
resolves https://github.com/stratum-mining/stratum/issues/1203

GitGab19 commented on protocols/v2/subprotocols/common-messages/src/setup_connection.rs:
```diff
@@ -16,38 +16,50 @@ use core::convert::TryInto;
+/// This is usually the first message sent by a downstream role on a newly opened connection,
+/// after completing the handshake process.
+///
+/// Downstream that do not wish to provide telemetry data to the upstream role **should** set
```
```suggestion
/// Downstreams that do not wish to provide telemetry data to the upstream role SHOULD set
```

jbesraa replied:
The "**word**"  makes the word **bold** on docs.rs, which I think makes more sense than all cap..

GitGab19 commented on protocols/v2/subprotocols/common-messages/src/setup_connection.rs:
```diff
@@ -16,38 +16,50 @@ use core::convert::TryInto;
-    /// The minimum protocol version the client supports (currently must be 2).
+    /// The minimum protocol version supported.
+    ///
+    /// must be set to 2.
```
```suggestion
    /// Currently it must be set to 2.
```

GitGab19 commented on protocols/v2/subprotocols/common-messages/src/setup_connection.rs:
```diff
@@ -16,38 +16,50 @@ use core::convert::TryInto;
-    /// The maximum protocol version the client supports (currently must be 2).
+    /// The maximum protocol version supported.
+    ///
+    /// must be set to 2.
```
```suggestion
    /// Currently it must be set to 2.
```

GitGab19 commented on protocols/v2/subprotocols/common-messages/src/setup_connection.rs:
```diff
@@ -16,38 +16,50 @@ use core::convert::TryInto;
     pub max_version: u16,
-    /// Flags indicating optional protocol features the client supports. Each
-    /// protocol from [`SetupConnection.protocol`] field has its own values/flags.
+    /// Indicating optional protocol features the downstream supports.
```
```suggestion
    /// Flags indicating optional protocol features supported by the downstream.
```

GitGab19 commented on protocols/v2/subprotocols/common-messages/src/setup_connection.rs:
```diff
@@ -235,41 +247,50 @@ impl<'a> From<SetupConnection<'a>> for CSetupConnection {
-/// ## SetupConnection.Success (Server -> Client)
-/// Response to [`SetupConnection`] message if the server accepts the connection. The client is
-/// required to verify the set of feature flags that the server supports and act accordingly.
+/// Used by an upstream role to accept a connection setup request from a downstream role.
```
```suggestion
/// Message used by an upstream role to accept a connection setup request from a downstream role.
```

GitGab19 commented on protocols/v2/subprotocols/common-messages/src/setup_connection.rs:
```diff
@@ -235,41 +247,50 @@ impl<'a> From<SetupConnection<'a>> for CSetupConnection {
     pub used_version: u16,
-    /// Flags indicating optional protocol features the server supports. Each
-    /// protocol from [`Protocol`] field has its own values/flags.
+    /// Indicating optional protocol features the upstream supports.
```
```suggestion
    /// Flags indicating optional protocol features supported by the upstream.
```

GitGab19 commented on protocols/v2/subprotocols/common-messages/src/setup_connection.rs:
```diff
@@ -235,41 +247,50 @@ impl<'a> From<SetupConnection<'a>> for CSetupConnection {
-/// cannot setup the connection) the server sends this message with a particular error code prior
-/// to closing the connection.
-/// In order to allow a client to determine the set of available features for a given server (e.g. for
+/// Used by an upstream role to reject a connection setup request from a downstream role.
```
```suggestion
/// Message used by an upstream role to reject a connection setup request from a downstream role.
```

Shourya742 commented on protocols/v2/subprotocols/common-messages/src/setup_connection.rs:
```diff
@@ -235,41 +247,50 @@ impl<'a> From<SetupConnection<'a>> for CSetupConnection {
+///
+/// This message is sent in response to a [`SetupConnection`] message.
+///
+/// The connection setup process could fail because of protocol version negotiation.  In order
```
```suggestion
/// The connection setup process could fail because of protocol version negotiation. In order
```

Shourya742 reviewed (changes requested):
@jbesraa, could you also update the `sv2.h` file to match the changes in the docs? I noticed a few public structures, methods, and enums missing documentation. Maybe add `#![deny(missing_docs)]` at the top level to catch those and see what's missing?

Also, these methods seem a bit off:

```rust
pub fn set_requires_standard_job(&mut self) {
    self.flags |= 0b_0000_0000_0000_0000_0000_0000_0000_0001;
}

pub fn set_async_job_negotiation(&mut self) {
    self.flags |= 0b_0000_0000_0000_0000_0000_0000_0000_0001;
}
```

Both are setting the LSB and `set_requires_standard_job` looks like it's only used in tests.

jbesraa commented:
> @jbesraa, could you also update the `sv2.h` file to match the changes in the docs? I noticed a few public structures, methods, and enums missing documentation. Maybe add `#![deny(missing_docs)]` at the top level to catch those and see what's missing?
> 
> Also, these methods seem a bit off:
> 
> ```rust
> pub fn set_requires_standard_job(&mut self) {
>     self.flags |= 0b_0000_0000_0000_0000_0000_0000_0000_0001;
> }
> 
> pub fn set_async_job_negotiation(&mut self) {
>     self.flags |= 0b_0000_0000_0000_0000_0000_0000_0000_0001;
> }
> ```
> 
> Both are setting the LSB and `set_requires_standard_job` looks like it's only used in tests.

Thanks for the review. What exactly should be changed in `sv2.h` ? is this the file in the `ffi` crate? 

In regards to the function that might be wrong/unused, I would leave that to a separate PR.

jbesraa commented:
@GitGab19 Thanks for the review. I believe I had addressed all of your comments. 

Some of the structs throw a warning for missing documentation for the `Deserliaze` proc-macro, I believe that will be fixed once the `binary_sv2` crate will be documented.

GitGab19 commented on protocols/v2/subprotocols/common-messages/src/lib.rs:
```diff
@@ -1,7 +1,15 @@
+#![warn(missing_docs)]
```
Why do we need this?

jbesraa replied:
This will tell us if we missed adding docs for any public facing API

GitGab19 commented on protocols/v2/subprotocols/common-messages/src/lib.rs:
```diff
@@ -1,7 +1,15 @@
 
-//! Common messages for [stratum v2][Sv2]
-//! The following protocol messages are common across all of the sv2 (sub)protocols.
+//! # Stratum V2 Common Messages Crate.
```
I would put this docs before `#![cfg_attr(feature = "no_std", no_std)]`

GitGab19 reviewed (commented):
LGTM. Left two minor things. For me we can open the PR review process to everyone now. What do you think @jbesraa ?

GitGab19 commented:
> @GitGab19 Thanks for the review. I believe I addressed all of your comments.
> 
> Some of the structs throw a warning(https://github.com/stratum-mining/stratum/actions/runs/11878855683/job/33100048008?pr=1242) for missing documentation for the `Deserliaze` proc-macro, I believe that will be fixed once the `binary_sv2` crate will be documented.

Why do you say it will be fixed when `binary_sv2` crate will be documented?

jbesraa commented:
> > @GitGab19 Thanks for the review. I believe I addressed all of your comments.
> > Some of the structs throw a warning(https://github.com/stratum-mining/stratum/actions/runs/11878855683/job/33100048008?pr=1242) for missing documentation for the `Deserliaze` proc-macro, I believe that will be fixed once the `binary_sv2` crate will be documented.
> 
> Why do you say it will be fixed when `binary_sv2` crate will be documented?

Because it is complaining about functions inside `Deserliaze` that are not documented and `Deserliaze` is coming from `binary_sv2`. We could remove the `warn(missing_docs)` also..

jbesraa commented:
> LGTM. Left two minor things. For me we can open the PR review process to everyone now. What do you think @jbesraa ?

sounds good

GitGab19 reviewed (commented):
LGTM. Let's open the review process to anyone to move on with this PR. I also saw that @Shourya742 already started to review it partially, don't know if he can complete it with a second one.

GitGab19 commented on protocols/v2/subprotocols/common-messages/README.md:
```diff
@@ -0,0 +1,21 @@
+[![license](https://img.shields.io/badge/license-MIT%2FApache--2.0-blue.svg)](https://github.com/stratum-mining/stratum/blob/main/LICENSE.md)
+[![codecov](https://codecov.io/gh/stratum-mining/stratum/branch/main/graph/badge.svg)](https://app.codecov.io/gh/stratum-mining/stratum/tree/main/protocols%2Fv2%2Fcommon_messages_sv2)
+
+`common_messages_sv2` is a Rust crate that implements a set of shared messages used by multiple Stratum V2 roles.
```
Isn't this a bit too much generic?

plebhash replied:
I think a more accurate description should highlight that these messages are shared across all Sv2 subprotocols.

plebhash commented on protocols/v2/subprotocols/common-messages/src/lib.rs:
```diff
@@ -1,7 +1,13 @@
+//! # Stratum V2 Common Messages Crate.
+//!
+//! The following crate defines a set of shared messages used across multiple Stratum V2 roles.
```
```suggestion
//! This crate defines a set of shared messages used across all Stratum V2 subprotocols.
```

plebhash commented on protocols/v2/subprotocols/common-messages/README.md:
```diff
@@ -0,0 +1,21 @@
+[![license](https://img.shields.io/badge/license-MIT%2FApache--2.0-blue.svg)](https://github.com/stratum-mining/stratum/blob/main/LICENSE.md)
+[![codecov](https://codecov.io/gh/stratum-mining/stratum/branch/main/graph/badge.svg)](https://app.codecov.io/gh/stratum-mining/stratum/tree/main/protocols%2Fv2%2Fcommon_messages_sv2)
+
+`common_messages_sv2` is a Rust crate that implements a set of shared messages used by multiple Stratum V2 roles.
```
```suggestion
`common_messages_sv2` is a Rust crate that implements a set of messages shared across all Stratum V2 subprotocols.
```

plebhash commented on protocols/v2/subprotocols/common-messages/src/setup_connection.rs:
```diff
@@ -137,49 +156,86 @@ impl<'decoder> SetupConnection<'decoder> {
     }
 }
 
+/// Checks if the flags include the Requires Standard Job flag.
```
```suggestion
/// Helper function to check if `SetupConnection` has `REQUIRES_STANDARD_JOBS` bit flag.
```

jbesraa replied:
Those functions dont know  anything about `SetupConnection` really, they just receive the flags

plebhash commented on protocols/v2/subprotocols/common-messages/src/setup_connection.rs:
```diff
@@ -137,49 +156,86 @@ impl<'decoder> SetupConnection<'decoder> {
     flag != 0
 }
+
+/// Checks if the flags include the Version Rolling flag.
```
```suggestion
/// Helper function to check if `SetupConnection` has `REQUIRES_VERSION_ROLLING` bit flag.
```

jbesraa replied:
Those functions dont know  anything about `SetupConnection` really, they just receive the flags

plebhash commented on protocols/v2/subprotocols/common-messages/src/setup_connection.rs:
```diff
@@ -137,49 +156,86 @@ impl<'decoder> SetupConnection<'decoder> {
     flag != 0
 }
+
+/// Checks if the flags include the Work Selection flag.
```
```suggestion
/// Helper function to check if [`SetupConnection`] has `REQUIRES_WORK_SELECTION ` bit flag.
```

jbesraa replied:
Those functions dont know  anything about `SetupConnection` really, they just receive the flags

plebhash commented on protocols/v2/subprotocols/common-messages/src/setup_connection.rs:
```diff
@@ -137,49 +156,86 @@ impl<'decoder> SetupConnection<'decoder> {
+    ///
+    /// Each [`SetupConnection::protocol`] value has it's own flags.
     pub flags: u32,
+    /// ASCII representation of the connection hostname or IP address.
     pub endpoint_host: CVec,
+    /// Connection port value.
     pub endpoint_port: u16,
+    /// ASCII representation of the device vendor name.
     pub vendor: CVec,
+    /// ASCII representation of the device hardware version.
     pub hardware_version: CVec,
+    /// ASCII representation of the device firmware version.
     pub firmware: CVec,
+    /// ASCII representation of the device identifier.
```
do we know for sure this is ASCII? where did you get this info?

plebhash commented on protocols/v2/subprotocols/common-messages/src/setup_connection.rs:
```diff
@@ -237,49 +293,58 @@ impl<'a> From<SetupConnection<'a>> for CSetupConnection {
 #[derive(Serialize, Deserialize, Debug, Clone, PartialEq, Eq)]
 pub struct SetupConnectionError<'decoder> {
-    /// Flags indicating features causing an error.
+    /// Unsupported feature flags.
+    ///
+    /// If set to 0, then this error is not due to unsupported flags.
```
semantics of this field should be primarily conditioned to `error_code`

```suggestion
    /// In case `error_code` is `unsupported-feature-flags`, this field is used to indicate which flag is causing an error
```

plebhash commented on protocols/v2/subprotocols/common-messages/src/setup_connection.rs:
```diff
@@ -237,49 +293,58 @@ impl<'a> From<SetupConnection<'a>> for CSetupConnection {
 #[repr(C)]
 #[cfg(not(feature = "with_serde"))]
 #[derive(Debug, Clone)]
+/// C representation of [`SetupConnectionSuccess`]
```
```suggestion
/// C representation of [`SetupConnectionError`]
```

Shourya742 commented:
> > @jbesraa, could you also update the `sv2.h` file to match the changes in the docs? I noticed a few public structures, methods, and enums missing documentation. Maybe add `#![deny(missing_docs)]` at the top level to catch those and see what's missing?
> > Also, these methods seem a bit off:
> > ```rust
> > pub fn set_requires_standard_job(&mut self) {
> >     self.flags |= 0b_0000_0000_0000_0000_0000_0000_0000_0001;
> > }
> > 
> > pub fn set_async_job_negotiation(&mut self) {
> >     self.flags |= 0b_0000_0000_0000_0000_0000_0000_0000_0001;
> > }
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Both are setting the LSB and `set_requires_standard_job` looks like it's only used in tests.
> 
> Thanks for the review. What exactly should be changed in `sv2.h` ? is this the file in the `ffi` crate?
> 
> In regards to the function that might be wrong/unused, I would leave that to a separate PR.

We can track the methods mentioned in the refactor issue for this crate to ensure they aren't forgotten. Regarding the `sv2.h` file, you need to run the `build_header.sh` script to generate an updated `sv2.h` file in the `scripts` folder. Then, copy its contents into the existing `sv2.h` file. Changes to the `sv2.h` file are due to newly added comments, which also ensure that the comments are rendered in the C++ header files.

Shourya742 commented on protocols/v2/subprotocols/common-messages/src/channel_endpoint_changed.rs:
```diff
@@ -6,16 +6,18 @@ use binary_sv2::{Deserialize, Serialize};
+/// When a downstream receives such a message, any extension state (including version and extension
+/// support) must be reset and renegotiated.
 #[repr(C)]
 #[derive(Serialize, Deserialize, Debug, Copy, Clone, PartialEq, Eq)]
```
```suggestion
#[derive(Serialize, Deserialize, Debug, Clone, PartialEq)]
```
This isn't related to documentation, but either `Copy` or `Clone` can be removed, as both aren't necessary. Additionally, `Eq` is not required. This could be added to the refactoring tracker.

Shourya742 commented on protocols/v2/subprotocols/common-messages/src/setup_connection.rs:
```diff
@@ -16,46 +16,62 @@ use core::convert::TryInto;
+    /// The minimum protocol version supported.
+    ///
+    /// Currently must be set to 2.
     pub min_version: u16,
-    /// The maximum protocol version the client supports (currently must be 2).
+    /// The maximum protocol version supported.
+    ///
+    /// Currently must be set to 2.
```
Why default for these are set to 2?

jbesraa replied:
I guess we only have stratum v2

GitGab19 commented:
@jbesraa is it ready for a final round of review? I see you still need to address some comments from @Shourya742

jbesraa commented:
> @jbesraa is it ready for a final round of review? I see you still need to address some comments from @Shourya742

I still need to fix the header thing, but feel free to review. Addressed all comments.

GitGab19 reviewed (commented):
LGTM. Left only a single minor comment.

GitGab19 commented on protocols/v2/subprotocols/common-messages/README.md:
```diff
@@ -0,0 +1,28 @@
+This crate can be built with the following features:
+
+ - `std`: Enables support for standard library features.
+ - `serde`: Enables support for serialization and deserialization using Serde.
```
```suggestion
 - `serde`: Enables support for serialization and deserialization using Serde. Note that this feature flag is only used for the Message Generator, and deprecated for any other kind of usage. It will likely be fully deprecated in the future.
```

jbesraa replied:
Thanks!

GitGab19 commented on protocols/v2/subprotocols/common-messages/src/lib.rs:
```diff
@@ -1,7 +1,16 @@
+//! ## Build Options
+//! This crate can be built with the following features:
+//! - `std`: Enables support for standard library features.
+//! - `serde`: Enables support for serialization and deserialization using Serde.
```
```suggestion
//! - `serde`: Enables support for serialization and deserialization using Serde. Note that this feature flag is only used for the Message Generator, and deprecated for any other kind of usage. It will likely be fully deprecated in the future.
```

GitGab19 replied:
This one is needed, otherwise it won't be rendered on crate docs

GitGab19 reviewed (commented):
It's good to go for me. I just pinged you on a couple of comments you might have missed.

jbesraa commented:
@GitGab19 Addressed the above

GitGab19 commented:
@jbesraa there's the semver check which is failing, when you solve it, I will merge this one.

jbesraa commented:
> @jbesraa there's the semver check which is failing, when you solve it, I will merge this one.

No idea why would this actually fail. could you please rerun it?
